src/storage/lsm.rs

```diff
@@ -69,98 +53,158 @@ impl State {
         };
         Ok(ret)
     }
-}
 
-fn read_commit_log(mut file: File) -> Result<Memtable> {
-    let mut get_data = || -> Result<Vec<u8>> {
-        let mut buf=[0u8; size_of::<usize>()];
-        &file.read_exact(&mut buf)?;
-        let sz = usize::from_le_bytes(buf);
-    
-        let mut buf = vec![0u8; sz];
-        &file.read_exact(&mut buf)?;
-        Ok(buf)
-    };
-    let mut get_key_value = || -> Result<(Key, Value)> {
-        let key_bytes = get_data()?;
-        let value_bytes = get_data()?;
-        let key = Key(String::from_utf8(key_bytes)?);
-        let val = Value::Bytes(value_bytes);
-        Ok((key, val))
-    };
+    pub fn flush_memtable(&mut self) -> Result<()> {
+        let path = PathBuf::from("/tmp/pancake/sstables/1.data");
+        let sstable = write_sstable(&self.memtable, path)?;
+        self.sstables.push(sstable);
+        self.memtable.0.clear();
+        self.commit_log.take(); // Close the file.
+        self.commit_log = Some(
+            OpenOptions::new()
+                .create(true)
+                .write(true)
+                .append(true)
+                .open(COMMIT_LOG_PATH)?,
+        );
+
+        Ok(())
+    }
+}
 
+/// Read all key-value pairs.
+fn read_commit_log(path: &PathBuf) -> Memtable {
     let mut memtable = Memtable::default();
-    loop {
-        if let Ok((key, val)) = get_key_value() {
-            memtable.0.insert(key, val);
-        } else {
-            break;
+
+    let file_result = File::open(path);
+    if let Ok(mut file) = file_result {
+        let iter = serde::KeyValueIterator { file: &mut file };
```

Comment:
Yeah good idea. We can make the happy path explicit (either missing file or file read successfully) and fail for all other cases.

And indeed we're not handling any i/o error, so the only possible recovery is from deleting files. We should think about better ways to automatically handle errors. Simply skipping files would lead to data corruption.